src/hera/workflows/runner.py

```diff
@@ -122,7 +123,53 @@ def _map_keys(function: Callable, kwargs: dict) -> dict:
     return mapped_kwargs
 
 
-def _runner(entrypoint: str, kwargs_list: Any) -> str:
+def _save_outputs(function_results, output_destinations):
+    if not isinstance(function_results, tuple):
+        function_results = [function_results]
+    if len(function_results) != len(output_destinations):
+        raise ValueError("The number of outputs does not match the annotation")
+
+    output_dir = Path(
+        str(global_config.experimental_features["hera_output_directory"])
+        if global_config.experimental_features["hera_output_directory"]
+        else "hera/outputs"
+    )
+
+    for res, dest in zip(function_results, output_destinations):
+        if not isinstance(res, dest[0]):
+            raise ValueError(
+                f"The type of output `{dest[1].name}`, `{type(res)}` does not match the annotated type `{dest[0]}`"
+            )
+        if isinstance(dest[1], Parameter):
+            path = output_dir / f"parameters/{dest[1].name}"
+        elif isinstance(dest[1], Artifact):
+            if dest[1].path:
+                path = Path(dest[1].path) / dest[1].name
+            else:
+                path = output_dir / f"artifacts/{dest[1].name}"
+        _write_to_path(path, res)
+
+
+def _extract_output_annotations(function: Callable):
+    output = []
+    if get_origin(inspect.signature(function).return_annotation) is Annotated:
+        output.append(get_args(inspect.signature(function).return_annotation))
+    elif get_origin(inspect.signature(function).return_annotation) is tuple:
+        for annotation in get_args(inspect.signature(function).return_annotation):
+            output.append(get_args(annotation))
+
+    return output or None
+
+
+def _write_to_path(path: Path, file) -> None:
+    directory = path.parent
+    directory.mkdir(parents=True, exist_ok=True)
+    path.touch(exist_ok=True)
+    result = serialize(file)
+    path.write_text(result if result else "")
+
+
+def _runner(entrypoint: str, kwargs_list: Any) -> Any:
```

Comment:
Small - given the documentation `The result of the function as a string or `None` if the outputs are to be saved.` should the return type by `Optional[str]`?

Reply:
Mikolaj actually pointed out the doc string seemed wrong but I guess he forgot to update it 😆 The return type should be `Any` because we're returning the value of a user function and only serializing it in `_run`. Not sure you gain much from `Optional[Any]` as it's technically covered by just `Any` but it's an option (pun intended 😎)

Reply:
Indeed, the return type used to be `str` which I fixed but I forgot the docstring